File: aspNetCoreBackend1/Services/PostInfoFetchingService.cs
using System.Text;
using System.Text.Json;


namespace aspNetCoreBackend1.Services;


public class PostInfoFetchingService
{


    public async Task<object> checkIfUserIsAnAuthorAndAlsoGetEncryptionStatusOfPost(
        int authUserId, string overallPostId, HttpClient httpClientWithMutualTLS
    )
    {
        try
        {
            HttpRequestMessage request = new HttpRequestMessage(
                HttpMethod.Get,
                @$"http://34.111.89.101/api/Home-Page/expressJSBackend1/getAuthorsAndEncryptionStatusOfPost
                /{overallPostId}"
            );

            HttpResponseMessage response = await httpClientWithMutualTLS.SendAsync(request);
            
            if (!response.IsSuccessStatusCode)
            {
                if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
                {
                    return (
                        @"There doesn't currently exist a post with the provided overallPostId",
                        "NOT_FOUND"
                    );                    
                }
                return (
                    @"The expressJSBackend1 server had trouble getting the authors and encryption-status of the post.",
                    "BAD_GATEWAY"
                );
            }

            string stringifiedResponseData = await response.Content.ReadAsStringAsync();
            Dictionary<string, object>? parsedResponseData = JsonSerializer.Deserialize<Dictionary<string, object>>(
                stringifiedResponseData
            );

            int[] authorsOfPost = (int[]) parsedResponseData!["authors"];
            bool? isEncrypted = (bool) parsedResponseData!["isEncrypted"];

            return (
                authorsOfPost.Contains(authUserId),
                isEncrypted
            );
        }
        catch
        {
            return (
                @"There was trouble connecting to the expressJSBackend1 server to get the authors and encryption-status of the
                post.",
                "BAD_GATEWAY"
            ); 
        }
    }


    public async Task<object> getPostEncryptionStatusIfUserHasAccessToPost(
        int? authUserId, string overallPostId, HttpClient httpClientWithMutualTLS
    )
    {  
        int[] authorsOfPost = [];
        bool isEncrypted = false;
        try
        {
            HttpRequestMessage request = new HttpRequestMessage(
                HttpMethod.Get,
                @$"http://34.111.89.101/api/Home-Page/expressJSBackend1/getAuthorsAndEncryptionStatusOfPost
                /{overallPostId}"
            );


            HttpResponseMessage response = await httpClientWithMutualTLS.SendAsync(request);
            

            if (!response.IsSuccessStatusCode)
            {
                if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
                {
                    return (
                        @"There doesn't currently exist a post with the overallPostId that you provided.",
                        "NOT_FOUND"
                    );                    
                }
                return (
                    @"The expressJSBackend1 server had trouble getting the authors and encryption-status of the post.",
                    "BAD_GATEWAY"
                ); 
            }

            string stringifiedResponseData = await response.Content.ReadAsStringAsync();
            Dictionary<string, object>? parsedResponseData = JsonSerializer.Deserialize<Dictionary<string, object>>(
                stringifiedResponseData
            );

            authorsOfPost = (int[]) parsedResponseData!["authorsOfPost"];
            isEncrypted = (bool) parsedResponseData["isEncrypted"];

            foreach(int author in authorsOfPost)
            {
                if (author == authUserId)
                {
                    return isEncrypted;
                }
            }
        }
        catch
        {
            return (
                @"There was trouble connecting to the expressJSBackend1 server to get the authors and encryption-
                status of the post.",
                "BAD_GATEWAY"
            ); 
        }

        if (isEncrypted)
        {
            if (authUserId == null)
            {
               return (
                    @"As an anonymous guest, you do not have access to this private-post or any of the encrypted data
                    associated with it.",
                    "UNAUTHORIZED"
                );  
            } 

            try
            {
                HttpRequestMessage request1 = new HttpRequestMessage(
                    HttpMethod.Post,
                    $"http://34.111.89.101/api/Home-Page/djangoBackend2/graphql"
                );

                request1.Content = new StringContent(
                    JsonSerializer.Serialize(new
                    {
                        query = @"query ($authUserId: Int!, $userIds: [Int!]!) {
                            checkIfUserFollowsAtLeastOneInList(authUserId: $authUserId, userIds: $userIds)
                        }",
                        variables = new
                        {
                            authUserId,
                            userIds = authorsOfPost
                        }
                    }),
                    Encoding.UTF8,
                    "application/json"
                );

                HttpResponseMessage response1 = await httpClientWithMutualTLS.SendAsync(request1);
                
                if (!response1.IsSuccessStatusCode)
                {
                    return (
                        @"The djangoBackend2 server had trouble verifying whether or not you follow at-least one of the
                        authors of this private-post.",
                        "BAD_GATEWAY"
                    ); 
                }

                string stringifiedDataForResponse1 = await response1.Content.ReadAsStringAsync();
                Dictionary<string, Dictionary<string, bool>>? parsedDataForResponse1 =  JsonSerializer.Deserialize<Dictionary<
                string, Dictionary<string, bool>>>(
                    stringifiedDataForResponse1
                );

                bool userFollowsAtLeastOneAuthor = parsedDataForResponse1!["data"]["checkIfUserFollowsAtLeastOneInList"];
                if (!userFollowsAtLeastOneAuthor) {
                    return (
                        @"You do not have access to any of the encrypted-data of this post since you do not follow
                        at-least one of its authors.",
                        "UNAUTHORIZED"
                    ); 
                }
            }
            catch
            {
                return (
                    @"There was trouble connecting to the djangoBackend2 server to verify whether or not you follow at-least 
                    one of the authors of this private-post.",
                    "BAD_GATEWAY"
                ); 
            }  
        }
        else
        {
            if (authUserId != null)
            {
                try
                {
                    HttpRequestMessage request2 = new HttpRequestMessage(
                        HttpMethod.Post,
                        @$"http://34.111.89.101/api/Home-Page/djangoBackend2/isEachUserInListInTheBlockingsOfAuthUser
                        /{authUserId}"
                    );

                    request2.Content = new StringContent(
                        JsonSerializer.Serialize(
                            new {
                                listOfUsers = authorsOfPost
                            }
                        ),
                        Encoding.UTF8,
                        "application/json"
                    );

                    HttpResponseMessage response2 = await httpClientWithMutualTLS.SendAsync(request2);

                    if (!response2.IsSuccessStatusCode)
                    {
                        return (
                            @"The djangoBackend2 server had trouble checking whether or not
                            each of the authors of this post either block you or are blocked by you.",
                            "BAD_GATEWAY"
                        ); 
                    }

                    string stringifiedResponse2Data = await response2.Content.ReadAsStringAsync();
                    bool? eachPostAuthorIsInAuthUserBlockings = JsonSerializer.Deserialize<bool>(
                        stringifiedResponse2Data
                    );
                    if (eachPostAuthorIsInAuthUserBlockings == true)
                    {
                        return (
                            "You are trying to access the data of a post that does not exist.",
                            "NOT_FOUND"
                        );
                    }
                }
                catch
                {
                    return (
                        @"There was trouble connecting to the djangoBackend2 server to check whether or not
                        each of the authors of this unencrypted post either block you or are blocked by you.",
                        "BAD_GATEWAY"
                    );
                }
            }
        }

        return isEncrypted;
    }

    public async Task<object> getAuthorsAndEncryptionStatusOfPostAndFollowingsAndBlockingsOfUser(
        int? authUserId, string overallPostId, HttpClient httpClientWithMutualTLS
    )
    {
        int[] authorsOfPost = [];
        bool isEncrypted = false;
        HashSet<int> setOfAuthUserFollowings = new HashSet<int>();
        HashSet<int> setOfAuthUserBlockings = new HashSet<int>();

        try
        {
            HttpRequestMessage request = new HttpRequestMessage(
                HttpMethod.Get,
                @$"http://34.111.89.101/api/Home-Page/expressJSBackend1/getAuthorsAndEncryptionStatusOfPost/{overallPostId}"
            );


            HttpResponseMessage response = await httpClientWithMutualTLS.SendAsync(request);
            

            if (!response.IsSuccessStatusCode)
            {
                if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
                {
                    return (
                        @"There doesn't currently exist a post with the overallPostId that
                        you provided.",
                        "NOT_FOUND"
                    );
                    
                }
                return (
                    @"The expressJSBackend1 server had trouble getting the authors and encryption-status of the post.",
                    "BAD_GATEWAY"
                ); 
            }

            string stringifiedResponseData = await response.Content.ReadAsStringAsync();
            Dictionary<string, object>? parsedResponseData = JsonSerializer.Deserialize<Dictionary<string, object>>(
                stringifiedResponseData
            );

            authorsOfPost = (int[]) parsedResponseData!["authorsOfPost"];
            isEncrypted = (bool) parsedResponseData["isEncrypted"];
        }
        catch
        {
            return (
                @"There was trouble connecting to the expressJSBackend1 server to get the authors and encryption-
                status of the post.",
                "BAD_GATEWAY"
            ); 
        }

        if (authUserId == -1)
        {
            authUserId = null;
        }

        if (authUserId != null)
        {
            try
            {
                HttpRequestMessage request1 = new HttpRequestMessage(
                    HttpMethod.Post,
                    $"http://34.111.89.101/api/Home-Page/djangoBackend2/getFollowingsAndBlockingsOfUser/{authUserId}"
                );

                HttpResponseMessage response1 = await httpClientWithMutualTLS.SendAsync(request1);
                
                if (!response1.IsSuccessStatusCode)
                {
                    return (
                        @"The djangoBackend2 server had trouble getting the people that you follow, as well
                        as the people who either block you or are blocked by you.",
                        "BAD_GATEWAY"
                    ); 
                }

                string stringifiedDataForResponse1 = await response1.Content.ReadAsStringAsync();
                Dictionary<string, object>? parsedDataForResponse1 = JsonSerializer.Deserialize<Dictionary<string, object>>(
                    stringifiedDataForResponse1
                );
                setOfAuthUserFollowings = new HashSet<int>((int[]) parsedDataForResponse1!["followings"]);
                setOfAuthUserBlockings = new HashSet<int>((int[]) parsedDataForResponse1!["blockings"]);
            }
            catch
            {
                return (
                    @"There was trouble connecting to the djangoBackend2 server to get the people that you follow, as well
                    as the people who either block you or are blocked by you.",
                    "BAD_GATEWAY"
                ); 
            }
        }


        return new Dictionary<string, object>
        {
            { "authorsOfPost", authorsOfPost },
            { "isEncrypted", isEncrypted },
            { "setOfAuthUserFollowings", setOfAuthUserFollowings },
            { "setOfAuthUserBlockings", setOfAuthUserBlockings },
        };
    }


    public async Task<object> GetOverallPostIdsOfEachUserInList(
        HttpClient httpClientWithMutualTLS, int[] listOfUserIds, int postsFromAtMostThisManyMonthsAgo,
        bool allUsersArePublic
    )
    {
        try
        {
            HttpRequestMessage request = new HttpRequestMessage(
                HttpMethod.Post,
                @$"http://34.111.89.101/api/Home-Page/expressJSBackend1/getTheOverallPostIdsOfEachUserInList
                /{postsFromAtMostThisManyMonthsAgo}/{allUsersArePublic}"
            );

            request.Content = new StringContent(
                JsonSerializer.Serialize(
                    new {
                        listOfUserIds,
                    }
                ),
                Encoding.UTF8,
                "application/json"
            );

            HttpResponseMessage response = await httpClientWithMutualTLS.SendAsync(request);

            if (!response.IsSuccessStatusCode)
            {
                return (
                    @"The expressJSBackend1 server had trouble getting the posts from at-most 2 months ago of each of the
                    users in the list",
                    "BAD_GATEWAY"
                ); 
            }

            string stringifiedResponseData = await response.Content.ReadAsStringAsync();
            Dictionary<string, object> parsedResponseData = JsonSerializer.Deserialize<Dictionary<string, object>>(
                stringifiedResponseData
            )!;
            Dictionary<string, int[]> overallPostIdsAndTheirAuthors = (Dictionary<string, int[]>) parsedResponseData[
                "overallPostIdsAndTheirAuthors"
            ];

            return overallPostIdsAndTheirAuthors;
        }
        catch
        {
            return (
                @"There was trouble connecting to the expressJSBackend1 server to get the posts from at-most 2 months ago of
                each of the users in the list",
                "BAD_GATEWAY"
            );
        }
    }


    public async Task<object> GetNumPostViewsOfEachOverallPostIdInList(
        HttpClient httpClientWithMutualTLS, List<string> overallPostIds
    )
    {
        try
        {
            HttpRequestMessage request = new HttpRequestMessage(
                HttpMethod.Post,
                "http://34.111.89.101/api/Home-Page/springBootBackend2/getNumPostViewsOfEachOverallPostIdInList"
            );

            request.Content = new StringContent(
                JsonSerializer.Serialize(
                    new {
                        overallPostIds
                    }
                ),
                Encoding.UTF8,
                "application/json"
            );

            HttpResponseMessage response = await httpClientWithMutualTLS.SendAsync(request);

            if (!response.IsSuccessStatusCode)
            {
                return (
                    @"The springBootBackend2 server had trouble getting the numPostViews received by the posts in the 
                    list",
                    "BAD_GATEWAY"
                ); 
            }

            string stringifiedResponseData = await response.Content.ReadAsStringAsync();
            Dictionary<string, int>? overallPostIdsAndTheirNumViews = JsonSerializer.Deserialize<Dictionary<string, int>>(
                stringifiedResponseData
            );

            return overallPostIdsAndTheirNumViews!;
        }
        catch
        {
            return (
                @"There was trouble connecting to the springBootBackend2 server to get the numPostViews received by the posts
                in the list",
                "BAD_GATEWAY"
            );
        }
    }


    public async Task<object> GetOverallPostIdsOfEachSponsoredPostThatAuthUserCanView(
        HttpClient httpClientWithMutualTLS, List<int> authUserFollowings, List<int> authUserBlockings,
        int postsFromAtMostThisManyMonthsAgo
    )
    {
        try
        {
            HttpRequestMessage request = new HttpRequestMessage(
                HttpMethod.Post,
                @$"http://34.111.89.101/api/Home-Page/expressJSBackend1
                /getTheOverallPostIdsOfEverySponsoredPostThatAuthUserCanView/{postsFromAtMostThisManyMonthsAgo}"
            );

            request.Content = new StringContent(
                JsonSerializer.Serialize(
                    new {
                        authUserFollowings,
                        authUserBlockings,
                    }
                ),
                Encoding.UTF8,
                "application/json"
            );

            HttpResponseMessage response = await httpClientWithMutualTLS.SendAsync(request);

            if (!response.IsSuccessStatusCode)
            {
                return (
                    @"The expressJSBackend1 server had trouble getting the sponsored posts from at-most 2 months ago that are
                    visible to the authUser",
                    "BAD_GATEWAY"
                ); 
            }

            string stringifiedResponseData = await response.Content.ReadAsStringAsync();
            Dictionary<string, object> parsedResponseData = JsonSerializer.Deserialize<Dictionary<string, object>>(
                stringifiedResponseData
            )!;
            Dictionary<string, int[]> overallPostIdsAndTheirAuthors = (Dictionary<string, int[]>) parsedResponseData[
                "overallPostIdsAndTheirAuthors"
            ];

            return overallPostIdsAndTheirAuthors;
        }
        catch
        {
            return (
                @"There was trouble connecting to the expressJSBackend1 server to get the sponsored posts from at-most 2
                months ago that are visible to the authUser",
                "BAD_GATEWAY"
            );
        }
    }


    public async Task<object> GetNumAdLinkClicksOfEachSponsoredOverallPostIdInList(
        HttpClient httpClientWithMutualTLS, List<string> sponsoredOverallPostIds
    )
    {
        try
        {
            HttpRequestMessage request = new HttpRequestMessage(
                HttpMethod.Post,
                "http://34.111.89.101/api/Home-Page/springBootBackend2/getNumAdLinkClicksOfEachSponsoredOverallPostIdInList"
            );

            request.Content = new StringContent(
                JsonSerializer.Serialize(
                    new {
                        sponsoredOverallPostIds
                    }
                ),
                Encoding.UTF8,
                "application/json"
            );

            HttpResponseMessage response = await httpClientWithMutualTLS.SendAsync(request);

            if (!response.IsSuccessStatusCode)
            {
                return (
                    @"The springBootBackend2 server had trouble getting the numAdLinkClicks received by the sponsored posts
                    of the list",
                    "BAD_GATEWAY"
                ); 
            }

            string stringifiedResponseData = await response.Content.ReadAsStringAsync();
            Dictionary<string, int> overallPostIdsAndTheirAdLinkClicks = JsonSerializer.Deserialize<Dictionary<string, int>>(
                stringifiedResponseData
            )!;

            return overallPostIdsAndTheirAdLinkClicks;
        }
        catch
        {
            return (
                @"There was trouble connecting to the springBootBackend2 server to get the numAdLinkClicks received by the
                sponsored posts in the list",
                "BAD_GATEWAY"
            );
        }
    }


    public async Task<object> GetNumPostViewsByAuthUserForEachOverallPostIdInList(
        HttpClient httpClientWithMutualTLS, List<string> overallPostIds, int authUserId
    )
    {
        try
        {
            HttpRequestMessage request = new HttpRequestMessage(
                HttpMethod.Post,
                @$"http://34.111.89.101/api/Home-Page/springBootBackend2
                /getNumPostViewsByAuthUserForEachOverallPostIdInList/{authUserId}"
            );

            request.Content = new StringContent(
                JsonSerializer.Serialize(
                    new {
                        overallPostIds
                    }
                ),
                Encoding.UTF8,
                "application/json"
            );

            HttpResponseMessage response = await httpClientWithMutualTLS.SendAsync(request);

            if (!response.IsSuccessStatusCode)
            {
                return (
                    @"The springBootBackend2 server had trouble getting the numPostViews by the authUser for each post in the 
                    list",
                    "BAD_GATEWAY"
                ); 
            }

            string stringifiedResponseData = await response.Content.ReadAsStringAsync();
            Dictionary<string, int>? overallPostIdsAndTheirNumViews = JsonSerializer.Deserialize<Dictionary<string, int>>(
                stringifiedResponseData
            );

            return overallPostIdsAndTheirNumViews!;
        }
        catch
        {
            return (
                @"There was trouble connecting to the springBootBackend2 server to get the numPostViews by the authUser for
                each post in the list",
                "BAD_GATEWAY"
            );
        }
    }

    public async Task<object> GetNumAdLinkClicksByAuthUserForEachSponsoredOverallPostIdInList(
        HttpClient httpClientWithMutualTLS, List<string> sponsoredOverallPostIds, int authUserId
    )
    {
        try
        {
            HttpRequestMessage request = new HttpRequestMessage(
                HttpMethod.Post,
                @$"http://34.111.89.101/api/Home-Page/springBootBackend2
                /GetNumAdLinkClicksByAuthUserForEachSponsoredOverallPostIdInList/{authUserId}"
            );

            request.Content = new StringContent(
                JsonSerializer.Serialize(
                    new {
                        sponsoredOverallPostIds
                    }
                ),
                Encoding.UTF8,
                "application/json"
            );

            HttpResponseMessage response = await httpClientWithMutualTLS.SendAsync(request);

            if (!response.IsSuccessStatusCode)
            {
                return (
                    @"The springBootBackend2 server had trouble getting the numAdLinkClicks by the authUser for
                    each sponsored post in the list",
                    "BAD_GATEWAY"
                ); 
            }

            string stringifiedResponseData = await response.Content.ReadAsStringAsync();
            Dictionary<string, int> overallPostIdsAndTheirAdLinkClicks = JsonSerializer.Deserialize<Dictionary<string, int>>(
                stringifiedResponseData
            )!;

            return overallPostIdsAndTheirAdLinkClicks;
        }
        catch
        {
            return (
                @"There was trouble connecting to the springBootBackend2 server to get the numAdLinkClicks by the authUser for
                each sponsored post in the list",
                "BAD_GATEWAY"
            );
        }
    }
}

File: springBootBackend2/src/main/java/com/megagram/springBootBackend2/services/PostInfoFetchingService.java
package com.megagram.springBootBackend2.services;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.springframework.stereotype.Service;

import com.fasterxml.jackson.databind.ObjectMapper;


@SuppressWarnings("unchecked")
@Service
public class PostInfoFetchingService {
    

    public PostInfoFetchingService() {}


    public Object checkIfPostIsSponsored(String overallPostId) {   
        try {
            URL url = new URL("http://34.111.89.101/api/Home-Page/expressJSBackend1/" +
            "checkIfPostIsSponsored/"+overallPostId);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");

            int responseCode = connection.getResponseCode();
            if (responseCode == HttpURLConnection.HTTP_NOT_FOUND) {
                return new String[]{
                    "There doesn\'t currently exist a post with the overallPostId that you provided.",
                    "NOT_FOUND"
                };
            }

            if (responseCode == HttpURLConnection.HTTP_OK) {
                return new String[]{
                    "The expressJSBackend1 server had trouble checking if the post is sponsored",
                    "BAD_GATEWAY"
                };
            }

            BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            reader.close();

            String stringifiedResponseData = response.toString();
            Boolean postIsSponsored = Boolean.parseBoolean(stringifiedResponseData);
            return postIsSponsored;
        }
        catch (Exception e) {
            return new String[] {
                "There was trouble connecting to the expressJSBackend1 server to check if the post is " +
                "sponsored",
                "BAD_GATEWAY"
            };
        }
    }


    public Object checkIfAuthUserIsAnAuthorOfPost(int authUserId, String overallPostId) {
        try {
            URL url = new URL("http://34.111.89.101/api/Home-Page/expressJSBackend1/" +
            "getAuthorsAndEncryptionStatusOfPost/"+overallPostId);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");

            int responseCode = connection.getResponseCode();
            if (responseCode == HttpURLConnection.HTTP_NOT_FOUND) {
                return new String[]{
                    "There doesn\'t currently exist a post with the overallPostId that you provided.",
                    "NOT_FOUND"
                };
            }

            if (responseCode == HttpURLConnection.HTTP_OK) {
                return new String[]{
                    "The expressJSBackend1 server had trouble getting the authors of the post.",
                    "BAD_GATEWAY"
                };
            }

            BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            reader.close();

            String stringifiedResponseData = response.toString();
            ObjectMapper objectMapper = new ObjectMapper();
            HashMap<String, Object> parsedResponseData = objectMapper.readValue(
                stringifiedResponseData, HashMap.class
            );

            List<Integer> authorsOfPost = (List<Integer>) parsedResponseData.get("authorsOfPost");

            return authorsOfPost.contains(authUserId);
        }
        catch (Exception e) {
            return new String[]{
                "There was trouble connecting to the expressJSBackend1 server to get the authors of the post.",
                "BAD_GATEWAY"
            };
        }
    }


    public Object checkIfAuthUserHasAccessToPost(int authUserId, String overallPostId) {
        List<Integer> authorsOfPost = new ArrayList<Integer>();
        Boolean isEncrypted = false;

        try {
            URL url = new URL("http://34.111.89.101/api/Home-Page/expressJSBackend1/" +
            "getAuthorsAndEncryptionStatusOfPost/"+overallPostId);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");

            int responseCode = connection.getResponseCode();
            if (responseCode == HttpURLConnection.HTTP_NOT_FOUND) {
                return new String[]{
                    "There doesn\'t currently exist a post with the overallPostId that you provided.",
                    "NOT_FOUND"
                };
            }

            if (responseCode == HttpURLConnection.HTTP_OK) {
                return new String[]{
                    "The expressJSBackend1 server had trouble getting the authors of the post.",
                    "BAD_GATEWAY"
                };
            }

            BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            reader.close();

            String stringifiedResponseData = response.toString();
            ObjectMapper objectMapper = new ObjectMapper();
            HashMap<String, Object> parsedResponseData = objectMapper.readValue(
                stringifiedResponseData, HashMap.class
            );

            authorsOfPost = (ArrayList<Integer>) parsedResponseData.get("authorsOfPost");

            if(authorsOfPost.contains(authUserId)) {
                return true;
            }

            isEncrypted = (Boolean) parsedResponseData.get("isEncrypted");
        }
        catch (Exception e) {
            return new String[]{
                "There was trouble connecting to the expressJSBackend1 server to get the authors of the post.",
                "BAD_GATEWAY"
            };
        }

        if (isEncrypted) {
            try {
                URL url = new URL("http://34.111.89.101/api/Home-Page/djangoBackend2/graphql");
                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                connection.setRequestMethod("POST");
                connection.setRequestProperty("Content-Type", "application/json");
                connection.setDoOutput(true);

                String query = 
                    "query ($authUserId: Int!, $user_ids: [Int!]!) { " + 
                        "checkIfUserFollowsAtLeastOneInList(authUserId: $authUserId, user_ids: $user_ids)"
                   + "}"
                ;
                HashMap<String, Object> variables = new HashMap<String, Object>();
                variables.put("authUserId", authUserId);
                variables.put("user_ids", authorsOfPost);

                HashMap<String, Object> requestBody = new HashMap<String, Object>();
                requestBody.put("query", query);
                requestBody.put("variables", variables);

                ObjectMapper objectMapper = new ObjectMapper();
                String requestBodyAsJSONString = objectMapper.writeValueAsString(requestBody);

                try (OutputStream os = connection.getOutputStream()) {
                    byte[] input = requestBodyAsJSONString.getBytes("utf-8");
                    os.write(input, 0, input.length);
                }
    
                int responseCode = connection.getResponseCode();
                if (responseCode == HttpURLConnection.HTTP_NOT_FOUND) {
                    return new String[]{
                        "There doesn\'t currently exist a post with the overallPostId that you provided.",
                        "NOT_FOUND"
                    };
                }
    
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    return new String[]{
                        "The djangoBackend2 server had trouble checking whether or not you follow " +
                        "at-least one of the authors of this post",
                        "BAD_GATEWAY"
                    };
                }
    
                BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
                StringBuilder response = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    response.append(line);
                }
                reader.close();
    
                String stringifiedResponseData = response.toString();
                objectMapper = new ObjectMapper();
                HashMap<String, HashMap<String, Boolean>> parsedResponseData = objectMapper.readValue(
                    stringifiedResponseData, HashMap.class
                );
                
                Boolean authUserFollowsAtLeastOnePostAuthor = parsedResponseData
                    .get("data")
                    .get("checkIfUserFollowsAtLeastOneInList");
                
                if (!authUserFollowsAtLeastOnePostAuthor) {
                    return new String[]{
                        "You do not follow at-least one of the authors of this post",
                        "UNAUTHORIZED"
                    };
                }
                return true;
            }
            catch (Exception e) {
                return new String[]{
                    "There was trouble connecting to the djangoBackend2 server to check if you follow " +
                    "at-least one of the authors of this post",
                    "BAD_GATEWAY"
                };
            }
        }
        else {
            try {
                URL url = new URL("http://34.111.89.101/api/Home-Page/djangoBackend2/" +
                "isEachUserInListInTheBlockingsOfAuthUser/" + authUserId);
                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                connection.setRequestMethod("POST");
                connection.setRequestProperty("Content-Type", "application/json");
                connection.setDoOutput(true);

                HashMap<String, List<Integer>> requestBody = new HashMap<String, List<Integer>>();
                requestBody.put("user_ids", authorsOfPost);

                ObjectMapper objectMapper = new ObjectMapper();
                String requestBodyAsJSONString = objectMapper.writeValueAsString(requestBody);

                try (OutputStream os = connection.getOutputStream()) {
                    byte[] input = requestBodyAsJSONString.getBytes("utf-8");
                    os.write(input, 0, input.length);
                }
    
                int responseCode = connection.getResponseCode();
    
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    return new String[]{
                        "The djangoBackend2 server had trouble checking whether or not each author of " +
                        "this post is in your blockings.",
                        "BAD_GATEWAY"
                    };
                }
    
                BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
                StringBuilder response = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    response.append(line);
                }
                reader.close();
    
                String stringifiedResponseData = response.toString();
                Boolean eachPostAuthorIsInBlockingsOfAuthUser = Boolean.parseBoolean(
                    stringifiedResponseData
                );
                
                if (!eachPostAuthorIsInBlockingsOfAuthUser) {
                    return new String[]{
                        "There doesn\'t currently exist a post with the overallPostId that you provided.",
                        "NOT_FOUND"
                    };
                }
                return true;
            }
            catch (Exception e) {
                return new String[]{
                    "There was trouble connecting to the djangoBackend2 server to check whether or not " +
                    "each author of the post is in your blockings",
                    "BAD_GATEWAY"
                };
            }
        }
    }
}
